### PR TITLE
stake-pool: Assess fee as a percentage of rewards

### LINKED ISSUE #1521: stake-pool: Change fee model to happen on "update pool balance"
#### Problem

Taking all of the fee upfront on deposit is a simple model, but unnecessarily punishes users who deposit for a short time.  Also, it's totally different from the main staking program's reward model if assessing the commission every epoch.

#### Solution

Adopt a model similar to regular staking, and have the stake pool fee paid out as pool tokens minted every epoch during "update pool balance".  Also, update the documentation to reflect this change.

### PR DESCRIPTION
**Pull request by joncinque**
Fixes #1521 

Instead of minting pool tokens on deposit to the manager, we mint them every epoch as a percentage of rewards perceived.  If there were no rewards or the pool lost value due to slashing, no fee is minted.  

From the associated issue:
> Problem
>
> Taking all of the fee upfront on deposit is a simple model, but unnecessarily punishes users who deposit for a short time. Also, it's totally different from the main staking program's reward model if assessing the commission every epoch.
>
> Solution
>
> Adopt a model similar to regular staking, and have the stake pool fee paid out as pool tokens minted every epoch during "update pool balance". Also, update the documentation to reflect this change.

### CODE REVIEW

File: stake-pool/program/src/instruction.rs [hunk 1]
```diff
@@ -41,7 +42,7 @@ pub enum StakePoolInstruction {
     ///   8. `[]` Rent sysvar
     ///   9. `[]` Token program id
     Initialize {
-        /// Deposit fee assessed
+        /// Fee assessed as percentage of perceived rewards
         #[allow(dead_code)] // but it's not
         fee: Fee,
```
Comment (mvines) [inline stake-pool/program/src/instruction.rs:47]: Totally fine as a follow-up, but I think we'll want an instruction to allow the fee to be changed after initialization

**Review by mvines — APPROVED** [review]
lgtm!

File: stake-pool/program/src/instruction.rs (hunk 1 above)
Comment (joncinque) [reply stake-pool/program/src/instruction.rs:47]: Yeah definitely, I didn't want to do put too much into this PR

**Comment by luciotato** [comment]
You're welcome 😜

**Pull request merged**